File: low-level-design/design-patterns/golang/creational/factory-pattern/factory/factory.go
package factory
import "fmt"

type StorageFactory interface {
	createStorageClient(provider string) (StorageClient, error)
}


func NewCloudStorageFactory() *CloudStorageFactory {
	return &CloudStorageFactory{}
}

type CloudStorageFactory struct {
}

func (f *CloudStorageFactory) CreateStorageClient(provider string) (StorageClient, error) {
	switch provider {
	case "aws":
		return NewAWSStorageClient(), nil
	case "gcp":
		return NewGCPStorageClient(), nil
	case "azure":
		return NewAzureStorageClient(), nil
	default:
		return nil, fmt.Errorf("unsupported cloud provider: %s", provider)
	}
}

File: low-level-design/design-patterns/python/creational/factory-pattern/factory.py
from azure_storage_client import AzureStorageClient
from aws_storage_client import AWSStorageClient
from gcp_storage_client import GCPStorageClient
from abc   import ABC,abstractmethod

class IStorageFactory(ABC):
    @abstractmethod
    def create_storage_client(self):
        pass

class CloudStorageFactory(IStorageFactory):
    @staticmethod
    def create_storage_client( cloud_provider: str):
        if cloud_provider == "AWS":
            return AWSStorageClient()
        elif cloud_provider == "Azure":
            return AzureStorageClient()
        elif cloud_provider == "GCP":
            return GCPStorageClient()
        else:
            raise ValueError("Invalid cloud provider")
